[Aidil] Duke Increments

aidilfbk opened the pull request.

j-lum committed:
Add support for Gradle workflow

j-lum committed:
Add sample checkstyle configuration

Add toolVersion block in to Gradle code sample to prevent errors.

aidilfbk committed:
Implemented basic Greet, Echo of user input and Exit on "bye" input

Added UserInterface class to handle user input and pretty printing

aidilfbk committed:
Implemented Add and List tasks

aidilfbk committed:
Implemented Mark as Done

Added Task class

aidilfbk committed:
Added a convenience method to UserInterface class, createStringJoiner.

Returned StringJoiner instance uses the platform's newline separator and can be initialized with a string.

aidilfbk committed:
Implemented ToDos, Events, Deadlines

aidilfbk committed:
Implemented Handle Errors via exceptions

aidilfbk committed:
Added support for deleting tasks from the list

j-lum committed:
Change file mode on `gradle` to be executable

j-lum committed:
Merge pull request #12 from j-lum/gradle+x

Change file mode on `gradle` to be executable (#9)

j-lum committed:
Add configuration for console applications

Gradle defaults to an empty stdin which results in runtime exceptions
when attempting to read from `System.in`. Let's add some sensible
defaults for students who may still need to work with the standard
input stream.

aidilfbk committed:
Ignore data folder

aidilfbk committed:
Updated Task subclass constructors to support serialization/unserialization

Classes affected:
– Deadline
– Event
– Todo

aidilfbk committed:
Added support for serialization/unserialization via TaskSerializer class.

DukeException now supports Chained Exception via a new constructor. Used by TaskSerializer.java's FileIOException/wrapping of Java IOException.

aidilfbk committed:
Duke now loads Tasks from "./data/duke.txt" and terminates on error if necessary

aidilfbk committed:
Duke now saves Tasks to "./data/duke.txt" whenever the task list changes

aidilfbk committed:
Duke now understands dates and times, parsing for Deadlines and Events in the "d/M/yyyy HHmm" format and displaying them when needed in the same format.

aidilfbk committed:
Merge branch 'branch-Level-8'.
Updated TaskSerializer to support Task subtype serialization/unserialization of dates and times.

# Conflicts:
#	src/main/java/Duke.java

j-lum committed:
Merge pull request #13 from j-lum/javaexec

Add configuration for console applications

aidilfbk committed:
Fixed inheritance hierarchy for DoneTokenParseError and TypeTokenParseError exception classes

aidilfbk committed:
Preliminary work to enable Duke instances

aidilfbk committed:
Moved Duke.formatTask static method to UserInterface class

aidilfbk committed:
TaskSerializer instances can now be created to load and save Tasks to a specific file path

aidilfbk committed:
Added TaskList class

aidilfbk committed:
Rearchitectured Duke

Created Command interface and reimplemented Duke's commands as separate classes.
Created Parser class to parse user text input and route to the correct Command implementing class.
Moved and created user interface display elements in UserInterface for common tasks.

aidilfbk committed:
Renamed UserInterface class to Ui in accordance with CS2103T module requirements

aidilfbk committed:
Renamed TaskSerializer class to Storage in accordance with CS2103T module requirements

aidilfbk committed:
Added Gradle Support via merge branch 'gradle' into A-JUnit

aidilfbk committed:
Modified Gradle settings

aidilfbk committed:
Added checkstyle to Gradle configuration

aidilfbk committed:
Broke out Command implementing classes to their own separate file

aidilfbk committed:
Broke out DukeException subclasses classes to their own separate file

aidilfbk committed:
Broke out task serializer exceptions to their own separate files

aidilfbk committed:
Modified Gradle settings

aidilfbk committed:
Override equals() and hashCode() function on Deadline, Event and Task classes

aidilfbk committed:
Added test cases for Storage class

aidilfbk committed:
AddCommand instances now trigger arguments parsing on instance creation

Previously triggered on calling the execute() function

aidilfbk committed:
Added test cases for Parser class

aidilfbk committed:
Added JUnit parameterized tests dependency

aidilfbk committed:
Added test cases for Ui class

aidilfbk committed:
Added new test case for Parser class, testing for unknown command input

aidilfbk committed:
Added a suite of tests for Command subclasses

– AddDeadlineTaskCommand
– AddEventTaskCommand
– AddTodoTaskCommand
– DeleteCommand
– DoneCommand
– ExitCommand
– ListCommand

aidilfbk committed:
Fixed Ui.createStringJoiner bug where passing in null would initialize it with the string "null"

Detected by UiTest.displayTasks test case

aidilfbk committed:
AddDeadlineTaskCommand parsing is now more robust

Enabled two test cases on AddDeadlineTaskCommandTest
– Test case for an exception when the "/by" argument is missing
– Test case for an exception when "/by" is provided but the date time argument is missing

aidilfbk committed:
AddEventTaskCommand parsing is now more robust

Enabled two test cases on AddEventTaskCommandTest
– Test case for an exception when the "/at" argument is missing
– Test case for an exception when "/at" is provided but the date time argument is missing

aidilfbk committed:
Parser is now more robust and transparently rethrows a DukeException from Command implementing class constructors

aidilfbk committed:
Fixed test cases for DeleteCommand and DoneCommand when passed in non numeric input

aidilfbk committed:
Duke can now be packaged as a JAR file

aidilfbk committed:
Implemented a way to find a task by searching for a keyword

aidilfbk committed:
Added test cases for FindTaskCommand

aidilfbk committed:
Preliminary coding standard compliance changes

Done through IntelliJ reformat code

aidilfbk committed:
Renamed FileIOException class to FileIoException class for coding style compliance

aidilfbk committed:
Conformed AddCommand.instantiateTask's access modifier keyword position to JLS suggestions

aidilfbk committed:
Conformed any lines of code longer than 120 characters

aidilfbk committed:
Parser.parse now uses if/else for simple commands instead of switch-case due to code style requiring a "default" clause

aidilfbk committed:
Marked the variables containing expected results in a few test cases as final

aidilfbk committed:
Added JavaDocs to a few methods.

– Duke constructor
– Duke.run()
– Parser.parse()
– Task.getStatusIcon()
– Ui.displayWelcome()
– Ui.displayLoadingError()
– Ui.displayError()
– Ui.displayTasks()
– Ui.displaySuccessfullyDoneTask()
– Ui.displaySuccessfullyRemovedTask()
– Ui.displaySuccessfullyAddedTask()

aidilfbk committed:
Merge branch 'Level-9' into A-JavaDoc

aidilfbk committed:
Merge branch 'A-JavaDoc' into A-CodingStandard

# Conflicts:
#	src/main/java/Duke.java
#	src/main/java/Ui.java

aidilfbk committed:
Renamed test case methods to the expected featureUnderTest_testScenario_expectedBehavior() format

aidilfbk committed:
Conforming some test case method signatures to code length limit

aidilfbk committed:
Marked the variables containing expected results in a few test cases as final

aidilfbk committed:
Added JavaFX to gradle config

aidilfbk committed:
Added GUI to Duke

Not wired to Duke's functionality yet, currently just echos the input

aidilfbk committed:
Wired GUI to Duke

Hacky method by intercepting the output of the Ui instance

aidilfbk committed:
Rearchitecture Ui class for later GUI support

– Ui class now turned into abstract class
– Previous Ui class renamed to UiCli class extending the new abstract Ui class
– UiTest renamed to UiCliTest

aidilfbk committed:
Add UiJavaFx class. Duke can now run in either CLI or GUI mode.

UiJavaFx extends from Ui abstract class and buffers Duke output for later display using a MainWindow controller.
UiJavaFx allows for Duke to display a message to the user at any time, unlike previously where it can only display a message in response to a user command. Used for displaying messages on startup (e.g. a welcome message or an error message when failed to load/parse file storage).

liawsy commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,46 @@
+public class Deadline extends Task {
+    protected LocalDateTime deadline;
+
+    public Deadline(String description, LocalDateTime deadline) {
```
missing javadocs

liawsy commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    /**
+     * Getter for when this Event occurs.
```
Instead of writing getter, use the format specified [here under the section for javadocs.](https://nus-cs2103-ay1920s1.github.io/website/coding-standards/java/intermediate.html)

liawsy commented on src/main/java/Parser.java:
```diff
@@ -0,0 +1,51 @@
+                    if (exc.getCause() instanceof DukeException) {
+                        throw (DukeException) exc.getCause();
+                    } else {
+                        throw new UnknownCommandException("I'm sorry, but I don't know what that means :-(", exc);
```
maybe instead of repeating this, you could create a function under your exceptions to print the error message.

liawsy commented on src/main/java/Storage.java:
```diff
@@ -0,0 +1,271 @@
+                throw new MissingTokenParseError(String.format(
+                        "Event task expected 4 values, received %d instead.",
+                        tokens.length
+                ));
```
i think this part should be attached to the previous line for coding standards

liawsy commented on src/main/java/Ui.java:
```diff
@@ -0,0 +1,233 @@
+    }
+
+    /**
+     * A convenience method to create a {@code StringJoiner} using the system's newline separator.
```
doesn't comply with javadoc standards

liawsy commented on src/test/java/AddDeadlineTaskCommandTest.java:
```diff
@@ -0,0 +1,68 @@
+import java.time.format.DateTimeParseException;
+import java.util.List;
+
+class AddDeadlineTaskCommandTest extends CommandTest {
```
comprehensive test cases

liawsy reviewed (commented):
It was good that you modularised your code. Additionally, your test cases were also very comprehensive. 

Here's some additional feedback:
1) Some parts of your code that didn't comply to the coding standards, such as missing javadoc comments and some formatting issues. 
2) Some parts of your code could be simplified.
3) Some variable names could be less lengthy and more explanatory.

Otherwise, good attempt! :)

chen-xi-cx commented on src/main/java/AddDeadlineTaskCommand.java:
```diff
@@ -0,0 +1,20 @@
+    }
+
+    protected Task instantiateTask() {
+        final String[] deadlineArgs = command.split("(?:\\A| )/by\\b ?");
```
Not sure if I missed out anything you are trying to capture but would it be better to use command.split("/by") instead which enhances readability.

chen-xi-cx commented on src/main/java/AddDeadlineTaskCommand.java:
```diff
@@ -0,0 +1,20 @@
+            throw new EmptyTaskDescriptionException("The description of a deadline cannot be empty.");
+        }
+
+        final String deadlineDueText = deadlineArgs[1].stripLeading();
```
would deadlineArgs[1].trim() be a better option here since trailing whitespaces can lead to unparseable date too.

chen-xi-cx commented on src/main/java/Storage.java:
```diff
@@ -0,0 +1,271 @@
+    }
+
+    /**
+     * Parse the Tasks from the input, adding to the provided List of the Tasks.
```
Minor error. Should be "Parses"

chen-xi-cx reviewed (commented):
The code is well thought out and well-written. Ambiguous methods were all commented which enhance readability. Some general comments:
1. Since you have so many classes, you may want to use package for your code files for better organization.
2. You give intuitive name to your methods so it is easy to infer what the method is doing, but I think it is good to javadoc comment at least all public methods (even for the obvious ones) and leave no room for ambiguity.

aidilfbk committed:
Add assert on Storage class

Ensures no null values in Iterable<Task> in the serialize method parameter.

aidilfbk committed:
Package classes

aidilfbk committed:
Reuse javafx.scene.image.Image instance on JavaFX GUI

aidilfbk committed:
Cleanup JavaDocs

aidilfbk committed:
Simplify Parser.parse exception handling

aidilfbk committed:
Made AddCommand.dateTimeFormatter into a static property

aidilfbk committed:
Removed Ui class reference from Ui methods

aidilfbk committed:
Add Override annotation to Command-implementing classes

aidilfbk committed:
Cleanup JavaDocs

aidilfbk committed:
Fix JavaDoc comment for Ui.printBlock

aidilfbk committed:
Restrict access modifier for DialogBox methods getUserDialog and getDukeDialog

aidilfbk committed:
Modify checkstyle to enforce access modifiers

aidilfbk committed:
Add JavaDocs for Command.execute and Command.isExit methods

aidilfbk committed:
General code cleanup

aidilfbk committed:
Restrict access modifier to DialogBox class

aidilfbk committed:
Add JavaDoc for AddCommand.instantiateTask method

aidilfbk committed:
Restrict access to AddCommand.task property

aidilfbk committed:
Add comment to explain how Parser.parse works

aidilfbk committed:
Merge pull request #5 from aidilfbk/A-Assertions

Add assert on Storage class

aidilfbk committed:
Merge pull request #6 from aidilfbk/A-CodeQuality

Improve Code Quality

aidilfbk committed:
Refactor FindTaskCommand to be Predicate-based
– Maintain the way to find a task by searching for a keyword from Level-9.
– Add "/isdone (yes|no)" subcommand for "find" command with test cases.

aidilfbk committed:
Add "/before", "/on", "/after" subcommands to FindTaskCommand

aidilfbk committed:
Add "/type (todo|deadline|task)" subcommand to FindTaskCommand

aidilfbk committed:
Add user friendly error message for "/type (todo|deadline|task)" subcommand in FindTaskCommand

aidilfbk committed:
Add user friendly error message for "/isdone (yes|no)" subcommand in FindTaskCommand

aidilfbk committed:
Cleanup code on duke.command.find.predicates

aidilfbk committed:
Set theme jekyll-theme-architect

aidilfbk committed:
Add User Guide with screenshot

aidilfbk committed:
Merge Github Pages theme commit '885c73abfd0199b8f73cf39b84ec22c60100cad7'

aidilfbk committed:
Fix Storage bug where it fails to create the parent directories before writing the Tasks to file

aidilfbk committed:
Fix Gradle settings so it can package the Duke GUI into a jar

aidilfbk committed:
Add feature to User Guide and changed placeholder symbol to be curly brackets instead

aidilfbk committed:
Merge User Guide changes from branch 'A-UserGuide'

aidilfbk committed:
Add cross-platform dependences for JavaFX to work when packaged in a JAR

aidilfbk committed:
Fix formatting on User Guide
